File: orion_utils_py/orion_utils_py/happy_birthday.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int64, Float64MultiArray, String
import math
import time


class OrionCelebrator(Node):

    def __init__(self):
        super().__init__('orion_celebrator')

        # --- Publishers ---
        self.emotion_pub = self.create_publisher(Int64, '/emotion/int', 10)
        self.arm_right_pub = self.create_publisher(Float64MultiArray, '/simple_right_arm_controller/commands', 10)
        self.arm_left_pub = self.create_publisher(Float64MultiArray, '/simple_left_arm_controller/commands', 10)
        self.tts_pub = self.create_publisher(String, '/orion_response', 10)

        # --- Publish emotion once at start ---
        msg = Int64()
        msg.data = 1
        self.emotion_pub.publish(msg)
        self.get_logger().info('Emotion published: 1')

        # --- Timer for arm waving ---
        self.start_time = time.time()
        self.create_timer(0.1, self.arm_motion_callback)  # 10 Hz

        # --- Publish creative birthday message ---
        bday_msg = String()
        bday_msg.data = "¡Feliz cumpleaños! Que tus circuitos brillen más que mis LEDs, y que tengas un año lleno de aventuras épicas."
        self.tts_pub.publish(bday_msg)
        self.get_logger().info(f'TTS message published: {bday_msg.data}')

    def arm_motion_callback(self):
        """Make arms oscillate between -1 and 1 rad (sine wave)."""
        elapsed = time.time() - self.start_time
        angle = math.sin(elapsed * 2.0)

        # Right arm
        msg_right = Float64MultiArray()
        msg_right.data = [angle]
        self.arm_right_pub.publish(msg_right)

        # Left arm
        msg_left = Float64MultiArray()
        msg_left.data = [-angle]
        self.arm_left_pub.publish(msg_left)


def main(args=None):
    rclpy.init(args=args)
    node = OrionCelebrator()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: orion_utils_cpp/src/happy_birthday.cpp
/*
 * @file happy_birthday.cpp
 * @brief Execute a birthday greeting sequence with arm oscillation and TTS for ORION.
 */

#include <chrono>
#include <cmath>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float64_multi_array.hpp"
#include "std_msgs/msg/int64.hpp"
#include "std_msgs/msg/string.hpp"

/* Node that publishes a birthday greeting with arm oscillation and TTS for ORION. */
class OrionCelebrator : public rclcpp::Node
{
public:
  /* Initialize publishers, send greeting message, and start arm timer. */
  OrionCelebrator()
  : Node("orion_celebrator")
  {
    emotion_pub_ = create_publisher<std_msgs::msg::Int64>("/emotion/int", 10);
    arm_right_pub_ = create_publisher<std_msgs::msg::Float64MultiArray>(
      "/simple_right_arm_controller/commands", 10);
    arm_left_pub_ = create_publisher<std_msgs::msg::Float64MultiArray>(
      "/simple_left_arm_controller/commands", 10);
    tts_pub_ = create_publisher<std_msgs::msg::String>("/orion_response", 10);

    // Publish emotion once at start
    std_msgs::msg::Int64 emotion_msg;
    emotion_msg.data = 1;
    emotion_pub_->publish(emotion_msg);
    RCLCPP_INFO(get_logger(), "Emotion published: 1");

    // Publish creative birthday message
    std_msgs::msg::String bday_msg;
    bday_msg.data =
      "¡Feliz cumpleaños! Que tus circuitos brillen más que mis LEDs, "
      "y que tengas un año lleno de aventuras épicas.";
    tts_pub_->publish(bday_msg);
    RCLCPP_INFO(get_logger(), "TTS message published: %s", bday_msg.data.c_str());

    start_time_ = now();

    // 10 Hz timer for arm waving
    timer_ = create_wall_timer(
      std::chrono::milliseconds(100),
      [this]() {arm_motion_callback();});
  }

private:
  /* Oscillate both arms between -1 and 1 rad using a sine wave. */
  void arm_motion_callback()
  {
    double elapsed = (now() - start_time_).seconds();
    double angle = std::sin(elapsed * 2.0);

    std_msgs::msg::Float64MultiArray msg_right;
    msg_right.data = {angle};
    arm_right_pub_->publish(msg_right);

    std_msgs::msg::Float64MultiArray msg_left;
    msg_left.data = {-angle};
    arm_left_pub_->publish(msg_left);
  }

  rclcpp::Time start_time_;

  rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr emotion_pub_;
  rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr arm_right_pub_;
  rclcpp::Publisher<std_msgs::msg::Float64MultiArray>::SharedPtr arm_left_pub_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr tts_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<OrionCelebrator>();
  try {
    rclcpp::spin(node);
  } catch (const std::exception &) {
    // intentional interrupt
  }
  rclcpp::shutdown();
  return 0;
}
